pkg/sentry/kernel/msgqueue/msgqueue.go

```diff
@@ -0,0 +1,275 @@
+// or IDs.
+//
+// +stateify savable
+type Registry struct {
+	// userNS owning the IPC namespace this registry belongs to. Immutable.
+	userNS *auth.UserNamespace
+
+	// mu protects all the fields below.
+	mu sync.Mutex `state:"nosave"`
+
+	// queues is a map of IDs to message queues.
+	queues map[ID]*Queue
+
+	// keysToIDs maps a lookup key to an ID.
+	keysToIDs map[Key]ID
+
+	// lastIDUsed is used to find the next available ID for queue creation.
+	lastIDUsed ID
+}
+
+// NewRegistry returns a new Registry ready to be used.
+func NewRegistry(userNS *auth.UserNamespace) *Registry {
+	return &Registry{
+		userNS:    userNS,
+		queues:    make(map[ID]*Queue),
+		keysToIDs: make(map[Key]ID),
+	}
+}
+
+// Queue represents a message queue that can be used to send and receive
+// text messages, and perform several control operations.
+//
+// +stateify savable
+type Queue struct {
+	// registry is the registry owning this queue. Immutable.
+	registry *Registry
+
+	// id is a kernel identifier for the queue. Immutable.
+	id ID
+
+	// key is a user-provided identifier for the queue. Immutable.
+	key Key
+
+	// creator is the user who created the queue. Immutable.
+	creator fs.FileOwner
+
+	// mu protects all the fields below.
+	mu sync.Mutex `state:"nosave"`
+
+	// dead is set to true when a queue is removed from the registry and should
+	// not be used. Operations on the queue should check dead, and return
+	// syserror.EIDRM if set to true.
+	dead bool
+
+	// senders holds a queue of blocked message senders. Senders are notified
+	// when enough space is available in the queue to insert their message.
+	senders waiter.Queue
+
+	// receivers holds a queue of blocked receivers. Receivers are notified
+	// when a new message is inserted into the queue and can be received.
+	receivers waiter.Queue
+
+	// messages is a list of sent messages.
+	messages []Message
+
+	// owner is the current owner of the queue.
+	owner fs.FileOwner
+
+	// perms is the access permissions for the queue.
+	perms fs.FilePermissions
+
+	// sendTime is the last time a msgsnd was perfomed.
+	sendTime ktime.Time
+
+	// receiveTime is the last time a msgrcv was performed.
+	receiveTime ktime.Time
+
+	// changeTime is the last time the queue was modified using msgctl.
+	changeTime ktime.Time
+
+	// byteCount is the current number of message bytes in the queue.
+	byteCount uint64
+
+	// messageCount is the current number of messages in the queue.
+	messageCount uint64
+
+	// maxBytes is the maximum allowed number of bytes in the queue, and is also
+	// used as a limit for the number of total possible messages.
+	maxBytes uint64
+
+	// sendPID is the PID of the process that performed the last msgsnd.
+	sendPID int32
+
+	// receivePID is the PID of the process that performed the last msgrcv.
+	receivePID int32
+}
+
+// Message represents a user-given block of memory.
+//
+// +stateify savable
+type Message struct {
```

Comment:
I decided to update `Queue.messages` to be a `msgList` (`ilist`) rather than a slice, to avoid insertion and deletion overhead. To do that, `Message` now has a `msgEntry` embedded field, which we can't add to `MsgBuf`. But yeah, I agree that it was somewhat useless before.